---
title: "body"
output: knitrBootstrap::bootstrap_document
---


imported variables:
  outPath
  fwdPrimer
  revPrimer
  targetTaxa
  primerName
  blastDb loc
  taxonomydb loc

```{bash mkdirs, eval = TRUE, cache = TRUE}
mkdir $outDir
mkdir ${outDir}/taxaRestricted
mkdir ${outDir}/taxaRestricted/primerTreeObj
#mkdir ${outDir}/figures
#mkdir ${outDir}/sequences
#mkdir ${outDir}/sequences/trimmed
#mkdir ${outDir}/blast
#mkdir ${outDir}/blast/topParsed
#mkdir ${outDir}/blast/potentialHits
#mkdir misc
#mkdir ${outDir}/unrestrictedPrimerTree/
#mkdir ${outDir}/unrestrictedPrimerTree/primerTreeObj
#mkdir ${outDir}/unrestrictedPrimerTree/figures
#mkdir ${outDir}/unrestrictedPrimerTree/taxonomy
#mkdir ${outDir}/unrestrictedPrimerTree/sequences
#mkdir ${outDir}/primerDiffs
```

### Read in the input file with primers
columns are:
primerNameF primerF primerNameR primerR targetTaxa

```{r getPrimers, eval = TRUE}
primerDf <- read.delim(primerFile, header = FALSE, sep = "\t")
```



### Calculate amplicon length distribution 


### Trim primers off seq?


### Filter unique seqs

Keep unique sequences per species and blast those, need to average number of hits within a species prior to other analyses to avoid eg. human causing massive skew of the data. 
```{r uniqueSeqs, engine = 'bash', eval = FALSE}
for file in output/sequences/trimmed/*.fasta
do
  base=${file##*/}
  base=${base%sequences.fasta}
  perl misc/uniqueFastaBySpeciesSeq.pl \
      --tax output/taxonomy/${base}taxonomy.txt \
      --fasta ${file} \
    > output/sequences/unique/${base}sequences.fasta
done
```


## Blast sequences
```{r blast, engine = 'bash', eval = FALSE}

for file in output/sequences/unique/*sequences.fasta
do
  base=${file%sequences.fasta}

  /usr/local/packages/ncbi-blast-2.2.31+/bin/blastn \
    -task blastn \
    -negative_gilist misc/unculturedOrgs.gi \
    -db /export/databases/blast/nt \
    -query ${file} \
    -num_threads 40 \
    -outfmt "7 qseqid sgi evalue length qlen sstart send slen" \
    -max_hsps 1 \
    -max_target_seqs 10000 \
    > output/blast/${base##*/}blastResults.txt
done
```

### Get taxonomy of blast hits

#### Pull hit GI number out of the BLAST results

```{r getGi, engine = 'bash', eval = FALSE}
cat output/blast/*blastResults.txt | grep -v "#" | cut -f 2,2 | perl -pe 's/gi.//' | perl -pe 's/\|.+//' | sort | uniq > output/blast/gis.txt
```


#### Get taxonomy information from NCBI

```{r getTaxonomy, engine = 'bash', eval = FALSE}
rm output/blast/taxa/splitInput/splitTaxa_*
split -a 4 -d output/blast/gis.txt output/blast/taxa/splitInput/splitTaxa_

parallel -j 20 '/usr/local/packages/perl-5.22.1/bin/perl ~/SerreDLab-3/cannonm3/scripts/getTaxaV2.pl --input {} --ranks "superkingdom, kingdom, phylum, class, order, family, genus" > output/blast/taxa/splitOutput/{/}' ::: output/blast/taxa/splitInput/splitTaxa_*
  
head -n 1 output/blast/taxa/splitOutput/splitTaxa_0000 > output/blast/blastTaxa.txt
cat output/blast/taxa/splitOutput/splitTaxa_* | grep -v "superkingdom" | grep -v "NotFound" >> output/blast/blastTaxa.txt
```

### Use taxonomy and blast output to keep all equal top hits
output: queryGi, orders, families, genera, species, numOrders, numFamilies, numGenera, numSpecies

```{r parseTopBlastHits, engine = 'bash', eval = FALSE}
for file in output/blast/*blastResults.txt
do
  base=${file%_blastResults.txt}
  base=${base##*/}
  perl misc/findTopBlastHitsWithTaxa.pl --blastIn ${file} --taxa output/blast/blastTaxa.txt > output/blast/topParsed/${base}_blastParsed.txt
done
```

### Count hits per rank
Need to check if the hits have taxonomy at this rank
```{r countNumInRank, engine = 'bash', eval = FALSE}
# family
for file in output/blast/topParsed/*_blastParsed.txt
do
  base=${file%_blastParsed.txt}
  base=${base##*/}
  grep -v "query" ${file} | cut -f 1,10 | grep -v "NA" | sort | uniq | cut -f 1,1 | sort | uniq -c | perl -pe 's/^ +//' > output/blast/topHitSummary/family/${base}Counts.txt
done
# no family info for blasto 
rm output/blast/topHitSummary/family/Blastocystis18S_F_Blastocystis18S_RCounts.txt

# genus
for file in output/blast/topParsed/*_blastParsed.txt
do
  base=${file%_blastParsed.txt}
  base=${base##*/}
  grep -v "query" ${file} | cut -f 1,11 | grep -v "NA" | sort | uniq | cut -f 1,1 | sort | uniq -c | perl -pe 's/^ +//' > output/blast/topHitSummary/genus/${base}Counts.txt
done

# species
for file in output/blast/topParsed/*_blastParsed.txt
do
  base=${file%_blastParsed.txt}
  base=${base##*/}
  grep -v "query" ${file} | cut -f 1,12 | grep -v "NA" | sort | uniq | cut -f 1,1 | sort | uniq -c | perl -pe 's/^ +//' > output/blast/topHitSummary/species/${base}Counts.txt
done
```




```{r listHitSpecies, engine = 'bash', eval = FALSE}

################
######### Double check field cuts
################

#banned words: sp., uncultured, unidentified, cf., isolate, symbiont

grep Apicomplexa output/taxonomy/Apicomplexa18S_365_F_Apicomplexa18S_613_R_taxonomyFixed.txt | cut -f 3,3 | grep -v "sp\." | grep -v "unidentified" | grep -v "cf\." | grep -v "isolate" | grep -v "symbiont" | grep -v "^NA$" | sort | uniq > output/taxonomy/Apicomplexa18S_365_F_Apicomplexa18S_613_R_SpeciesList.txt
 
## parabasalia is distinct due to taxonomy
########################### check
grep -f parabasaliaGenera.txt output/taxonomy/Parabasalia18S_288_F_Parabasalia18S_654_R_taxonomyFixed.txt | cut -f 3,3 | grep -v "sp\." | grep -v "cf\." | grep -v "isolate" | grep -v "unidentified" | grep -v "symbiont" |  grep -v "^NA$" | sort | uniq > output/taxonomy/Parabasalia18S_288_F_Parabasalia18S_654_R_SpeciesList.txt

wc -l output/taxonomy/*SpeciesList.txt | perl -pe 's/^\s+//' > output/summaryTable/hitsSpeciesCount_SpeciesList.txt

```


## Find missed hits
banned words: sp., uncultured, unidentified, cf., isolate, symbiont
```{r parsePotentialHits, engine = 'bash', eval = FALSE}
# Apicomplexa
perl misc/getPotentialHits.pl --alignVar 10 --taxa output/blast/blastTaxa.txt --blast output/blast/Apicomplexa18S_365_F_Apicomplexa18S_613_R_blastResults.txt --primerF GACCTATCAGCTTTCGACGG --primerR CCCTCCAATTGWTACTCTGGR > output/blast/potentialHits/Apicomplexa18S_365_F_Apicomplexa18S_613_R_potential_hits.txt

# Parabasalia
perl misc/getPotentialHits.pl --alignVar 10 --taxa output/blast/blastTaxa.txt --blast output/blast/Parabasalia18S_288_F_Parabasalia18S_654_R_blastResults.txt --primerF TAGGCTATCACGGGTAACGG --primerR GCGTCCTGATTTGTTCACAG > output/blast/potentialHits/Parabasalia18S_288_F_Parabasalia18S_654_R_potential_hits.txt
```

### 

```{r countUniqueSpecies, engine = 'bash', eval = FALSE}
grep Apicomplexa output/blast/potentialHits/Apicomplexa18S_365_F_Apicomplexa18S_613_R_potential_hits.txt | cut -f 8,8 | grep -v species | sort | uniq | grep -v "^NA$" | perl -pe 's/_/ /' > output/blast/potentialHits/Apicomplexa18S_365_F_Apicomplexa18S_613_R_species_list.txt

wc -l output/blast/potentialHits/*_species_list.txt | perl -pe 's/^\s+//' > output/summaryTable/potentialHitsSpeciesCount_species_list.txt
```


### Count number of mismatches between primer and template

```{bash countPrimerDiffsForPaper, eval = FALSE}
perl misc/checkPrimerMismatches.pl --primerL GACCTATCAGCTTTCGACGG --primerR CCCTCCAATTGWTACTCTGGR --fasta output/sequences/Apicomplexa18S_365_F_Apicomplexa18S_613_R_sequences.fasta > output/primerDiffs/Apicomplexa18S_365_F_Apicomplexa18S_613_R_PrimerDiffs.txt

```

#### Plot primer diffs

```{r plotPrimerDiffsForPaper, eval = FALSE}
diffFiles <- list.files(path = 'output/primerDiffs/', pattern = "PrimerDiffs.txt", full.names = T)

pdf("output/figures/primerDiffs.pdf", width = 15, height = 8)
for(inFile in diffFiles) {
  baseName <- gsub(".+/", "", inFile)
  baseName <- gsub("_PrimerDiffs.txt", "", baseName)  
  diffCount <- read.delim(inFile, header = F)
  colnames(diffCount) <- c("gi", "RprimerMismatches", "RprimerMatches", "LprimerMismatches", "LprimerMatches")
  
  taxaDf <- read.delim(paste("output/taxonomy/", baseName, "_taxonomyFixed.txt", sep = ""), header = T)
  
  combinedDf <- merge(diffCount, taxaDf)
  combinedDf$count <- 1
  combinedDf <- subset(combinedDf, RprimerMismatches < 10 & LprimerMismatches < 10)

  for(taxon in c("order", "family", "genus")) {
    if(taxon %in% colnames(combinedDf)) {
      # print(ggplot(combinedDf, aes(x = get(taxon), y = count, fill = as.factor(RprimerMismatches))) + 
      #     geom_bar(stat = "identity", position = "fill") +
      #     ggtitle(baseName) +
      #     theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
      #     guides(fill = guide_legend(title = "Primer\nmismatches")) +
      #     xlab("") +
      #     ylab("Proportion")
      # )
      
      print(ggplot(combinedDf, aes(x = get(taxon), y = as.factor(RprimerMismatches))) +
          geom_bin2d() +
          ggtitle(paste(baseName, taxon, "Right primer", sep = "\n")) +
          theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
          guides(fill = guide_legend(title = "Primer\nmismatches")) +
          xlab("") +
          ylab("Proportion")
      )

      print(ggplot(combinedDf, aes(x = get(taxon), y = as.factor(LprimerMismatches))) +
          geom_bin2d() +
          ggtitle(paste(baseName, taxon, "Left primer", sep = "\n")) +
          theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
          guides(fill = guide_legend(title = "Primer\nmismatches")) +
          xlab("") +
          ylab("Proportion")
      )
      
      

    }
  }
}
dev.off()

```



## Make up summary table for paper
use:
Percent of species on target
  output/summaryTable/OnTargetSpeciesCount_SpeciesList.txt
  output/summaryTable/OffTargetSpeciesCount_SpeciesList.txt

Number of species found and in NCBI
  output/summaryTable/hitsSpeciesCount_SpeciesList.txt
  output/summaryTable/potentialHitsSpeciesCount_species_list.txt

Primer taxonomic specificity
  output/blast/topHitSummary/family/${base}Counts.txt
  output/blast/topHitSummary/genus/${base}Counts.txt
  output/blast/topHitSummary/species/${base}Counts.txt

```{r makingSummaryTable, eval = FALSE}
#############################################
### Number of species found and in NCBI total
foundCounts <- read.delim("output/summaryTable/hitsSpeciesCount_SpeciesList.txt", header = F, sep = " ")
colnames(foundCounts) <- c("Species found", "Primer")
foundCounts <- subset(foundCounts, Primer != "total")
foundCounts$Primer <- gsub(".+/", "", foundCounts$Primer)
foundCounts$Primer <- gsub("_SpeciesList.txt", "", foundCounts$Primer)

potentialCounts <- read.delim("output/summaryTable/potentialHitsSpeciesCount_species_list.txt", header = F, sep = " ")
colnames(potentialCounts) <- c("Species In NCBI", "Primer")
potentialCounts <- subset(potentialCounts, Primer != "total")
potentialCounts$Primer <- gsub(".+/", "", potentialCounts$Primer)
potentialCounts$Primer <- gsub("_species_list.txt", "", potentialCounts$Primer)

summaryTable <- merge(foundCounts, potentialCounts, all = T)

summaryTable$`Species In NCBI` <- paste(
  prettyNum(summaryTable$`Species In NCBI`, big.mark = ","), 
  " (", 
  round(100 * (summaryTable$`Species found` / summaryTable$`Species In NCBI`), digits = 1), 
  "%)", 
  sep = "")

summaryTable$`Species found` <- prettyNum(summaryTable$`Species found`, big.mark = ",")

rownames(summaryTable) <- summaryTable$Primer

##############################################
### Mean number of species matching each blast query and % single match
for(tLevel in c("family", "genus", "species")) {
  specificityFiles <- list.files(paste("output/blast/topHitSummary/", tLevel, "/", sep = ""), pattern = "Counts.txt", full.names = T)
  for(inFile in specificityFiles) {
    Primer <- gsub(".+/", "", inFile)
    Primer <- gsub("Counts.txt", "", Primer)
    
    dataDf <- read.delim(inFile, header = F, sep = " ")
    
    summaryTable[Primer, paste(tLevel, " mean", sep = "")] <- round(mean(dataDf$V1), digits = 1)
    
    summaryTable[Primer, paste(tLevel, " % single", sep = "")] <- paste(round(100 * (sum(dataDf$V1 == 1) / length(dataDf$V1)), digits = 1), "%", sep = "")
  }
}

###############################################
### % species on target
onTargetCounts <- read.delim("output/summaryTable/OnTargetSpeciesCount_SpeciesList.txt", header = F, sep = " ")
colnames(onTargetCounts) <- c("On target", "Primer")
onTargetCounts <- subset(onTargetCounts, Primer != "total")
onTargetCounts$Primer <- gsub(".+/", "", onTargetCounts$Primer)
onTargetCounts$Primer <- gsub("_SpeciesList.txt", "", onTargetCounts$Primer)
onTargetCounts$Primer <- gsub("On_", "", onTargetCounts$Primer)

summaryTable <- merge(summaryTable, onTargetCounts, all = T)

offTargetCounts <- read.delim("output/summaryTable/OffTargetSpeciesCount_SpeciesList.txt", header = F, sep = " ")
colnames(offTargetCounts) <- c("Off target", "Primer")
offTargetCounts <- subset(offTargetCounts, Primer != "total")
offTargetCounts$Primer <- gsub(".+/", "", offTargetCounts$Primer)
offTargetCounts$Primer <- gsub("_SpeciesList.txt", "", offTargetCounts$Primer)
offTargetCounts$Primer <- gsub("Off_", "", offTargetCounts$Primer)

summaryTable <- merge(summaryTable, offTargetCounts, all = T)

summaryTable$`% Species on-target` <- paste(round(100 * (summaryTable$`On target` / (summaryTable$`On target` + summaryTable$`Off target`)), digits = 1), "%", sep = "")

rownames(summaryTable) <- summaryTable$Primer

##############################################
### reorder rows
properOrder <- c("Amoebozoa_18S_2_rev", "Apicomplexa", "Apicomplexa18S_365_F_Apicomplexa18S_613_R", "Eimeriorina18S_302_F_Eimeriorina18S_730_R", "Plasmodium18S_883_F_Plasmodium18S_1126_R", "Blastocystis18S_F_Blastocystis18S_R", "Diplomonadida_768_F_Diplomonadida_1059_R", "Kinetoplastida_18S_4_Kinetoplastida_18S_4", "Microsporidia_18S_F_Microsporidia_18S_R", "Nematoda", "Spirurida18S_1435_F_Spirurida18S_1858_R", "Spirurida18S_F2_Spirurida18S_R2", "Trichocephalida18S2_F_Trichocephalida18S2_R", "Parabasalia18S_288_F_Parabasalia18S_654_R", "Platyhelminthes_18S3_F_Platyhelminthes_18S3_R")

summaryTable <- summaryTable[properOrder,]

colnames(summaryTable) <- gsub("family", "Family", colnames(summaryTable))
colnames(summaryTable) <- gsub("genus", "Genus", colnames(summaryTable))
colnames(summaryTable) <- gsub("species", "Species", colnames(summaryTable))

write.table(summaryTable, file = 'output/summaryTable/finalTable.txt', quote = F, sep = "\t", row.names = F, col.names = T, na = "")

```
